numpyro/util.py

```diff
@@ -331,48 +332,77 @@ def _body_fn(i, vals):
         val, collection, start_idx, thinning = vals
         val = body_fun(val)
         idx = (i - start_idx) // thinning
-        collection = cond(
-            idx >= 0,
-            collection,
-            lambda x: x.at[idx].set(ravel_pytree(transform(val))[0]),
-            collection,
-            identity,
-        )
+
+        def update_fn(collect_array, new_val):
+            return cond(
+                idx >= 0,
+                collect_array,
+                lambda x: x.at[idx].set(new_val),
+                collect_array,
+                identity,
+            )
+
+        def update_collection(collection, val):
+            return jax.tree.map(update_fn, collection, transform(val))
+
+        collection = jit(update_collection, donate_argnums=0)(collection, val)
+
         return val, collection, start_idx, thinning
 
-    collection = jnp.zeros(
-        (collection_size,) + init_val_flat.shape, dtype=init_val_flat.dtype
-    )
+    @partial(jit, donate_argnums=2)
+    @cached_by(fori_collect, body_fun, transform)
+    def _body_fn_wrap(i, val, collection, start_idx, thinning):
+        return _body_fn(i, (val, collection, start_idx, thinning))
+
+    def map_fn(x):
+        nx = jnp.asarray(x)
+        return jnp.zeros((collection_size, *nx.shape), dtype=nx.dtype) * nx[None, ...]
```

Comment:
The nx[None, ...] is forcing the array to become a tracer before it is donated in jax.jit.  Otherwise under vmap the buffer can't be donated since it is just length 1 in the vmapped dimension